pkg/providers/vault/client.go

```diff
@@ -24,17 +25,39 @@ import (
 	"net/http"
 	"time"
 
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/logger"
+
 	"github.com/edgexfoundry/go-mod-secrets/pkg"
 )
 
+const (
+	AuthTypeHeader = "X-Vault-Token"
+)
+
 // Client defines the behavior for interacting with the Vault REST secret key/value store via HTTP(S).
 type Client struct {
 	HttpConfig SecretConfig
 	HttpCaller Caller
+
+	// internal member variables
+	lc logger.LoggingClient
 }
 
+// package stateful map variable to handle the case of the same caller to have
+// multiple secret clients with potentially the same tokens while renewing token
+// in the background go-routine
+var vaultTokenToCancelFunc = make(map[string]context.CancelFunc)
+
 // NewSecretClient constructs a SecretClient which communicates with Vault via HTTP(S)
-func NewSecretClient(config SecretConfig) (pkg.SecretClient, error) {
+// lc is the Edgex customized logger
+// bkgCtx is the background context that can be used to cancel or cleanup the background process when it is no longer needed
+// errChan is the error channel to receive any error from the background go-routine calls
+func NewSecretClient(config SecretConfig, lc logger.LoggingClient, bkgCtx context.Context, errChan chan<- error) (pkg.SecretClient, error) {
```

Comment:
You require an error stream be passed in.  I understand this is how you're surfacing errors that occur in future asynchronous attempts to renew a token.  What exactly do you expect the caller of `NewSecretClient` to do with these errors?  

Is there some behavior that manifests itself in the existing API (i.e. `GetSecrets` and `StoreSecrets`) when a token was able to be renewed?  Could this error channel be eliminated (it is leaking implementation detail) in favor of providing an error when the API is called?

Reply:
> 
> 
> You require an error stream be passed in. I understand this is how you're surfacing errors that occur in future asynchronous attempts to renew a token. What exactly do you expect the caller of `NewSecretClient` to do with these errors?

It is really up to the caller to handle that. The background process does not pose a policy or action to do in that case.  Possible actions the caller can do are: 1) take the returned error from the stream and then log it to that the end user know the token cannot be renewed, OR 2) take the error and the caller can decide to use different token if there is such mechanism to get one from the caller side.  Again, the module should not restrict the caller how to handle the token renewal error and is up to the caller to handle it accordingly.

Reply:
> Is there some behavior that manifests itself in the existing API (i.e. `GetSecrets` and `StoreSecrets`) when a token was able to be renewed? Could this error channel be eliminated (it is leaking implementation detail) in favor of providing an error when the API is called?

Not I am aware of.  The existing APIs take the token and use it to authenticate with Vault but have no knowledge at all whether the token should be renewed or not.  Authentication with token is one thing, renewal of token is another thing.   We should not mix both together for the separation of duties concern.

Reply:
As the calling application, if I can't take a specific discrete action as a result of a token not being renewed, there's no real reason to asynchronously notify me that it hasn't.  It just complicates the implementation.  

Is there a discrete action the caller could take that I'm missing that would cause the token to be renewed? 

You could say "yeah, but you should be able to log those errors."  To which I'd respond that you're already having the caller provide an implementation that would allow you to log them yourself.  Which, I'd suggest, you should (if you're not already).

So if we fail to renew the token and subsequently call `GetSecrets` or `StoreSecrets`, those calls will return normally?  Or will they return an error?

From an internal implementation perspective, you're attempting to "separate concerns."  Unfortunately, you're leaking implementation detail to the caller that I can't see the caller caring about.  The caller doesn't see these as two separate concerns.  The caller only cares that it needs to construct a secret client and that future calls to `GetSecrets` returns secrets and future calls to `StoreSecrets` store them.

Reply:
You have good valid points.  But if token is expired, then you will not be able to make subsequently call `GetSecrets` or `StoreSecrets`. The Vault server will just respond with permission denied. In that case, I think you have to get a new token which is out of scope of this token refresh PR. It is the caller to decide what to do on that case.

Reply:
How about this:  If the token fails to renew, that doesn't mean that it's necessarily dead yet.  How about if the call fails reset the timer for one minute, and just run the clock out.  Eventually the client will get an error that it has to do something about once the token finally expires.

Reply:
Moreover, if we actually get a 403 error, that means the token is indeed dead and we can stop retrying.

Reply:
> 
> 
> How about this: If the token fails to renew, that doesn't mean that it's necessarily dead yet. How about if the call fails reset the timer for one minute, and just run the clock out, retrying every minute. Eventually the client will get an error that it has to do something about once the token finally expires.

That could alleviate the connection lost issue for the time being, but if you get a forbidden 403 from Vault's response, it is basically permission issue and that token is basically useless: you need a new token no matter what, I think.

Reply:
There's a larger decision to be made here -- what to do when a token expires.  I think #security needs to have a discussion (@mkbhanda, @tingyuz) to answer that question.

With the current PR implementation, you're basically tossing that decision over the wall to the caller (without explicitly calling out that you're doing so) for them to deal with.  Core and support services won't know what to do with this edge case.  It makes more sense to come up with a standard approach and implement it in a centralized place.  You can remain flexible via strategy pattern -- create a contract to call when this edge case occurs, the referential implementation to satisfy the contract, and pass the implementation into the appropriate receiver's factory function.  This allows for future change and augmentation without someone having to reopen the referential implementation.

`GetSecrets` and `StoreSecrets` will fail without a valid token.  I'm not sure that's accounted for in `go-mod-bootstrap`'s usage of `vault.NewSecretClient`.  Currently, we initialize the client once and call `GetSecrets` to get credentials once.  This is all during startup.  If a secret-related call fails, we stop the service.

Going forward where we're dealing with secrets that are stored and retrieved outside of the start up window, we'll need to ensure the caller has a strategy for dealing with retries.  It would be ideal for the caller remain ignorant of the details.  But as you said, if the token expires we're screwed -- the solution is beyond the scope of this PR.  Having said that, I'd hope there's a follow-on issue to cover that eventuality and some strategy has already been devised on how to deal with it.

Reply:
Updated to provide a referential call back mechanism from the caller side if token is somehow expired during the background renewal process...

Reply:
In particular, 
> You can remain flexible via strategy pattern -- create a contract to call when this edge case occurs, the referential implementation to satisfy the contract, 

Implemented constructor input argument with `TokenExpiredCallback` func to call back when the token renewal receives 403 forbidden.
```go

type TokenExpiredCallback func(expiredToken string) (replacementToken string, retry bool) 

```

> and pass the implementation into the appropriate receiver's factory function. This allows for future change and augmentation without someone having to reopen the referential implementation.

The caller can re-read the token from the disk to see if it was updated offline, and set the retry to be true while supplying a new token if any.

Reply:
Thank you.  This looks much better -- you've done what I've asked.  

I have a follow-on request that you create a new issue in go-mod-bootstrap to make the necessary changes to reflect the contract/implementation changes you made here along.  All of core's interaction with go-mod-secrets is encapsulated there.  

You could also implement the default callback to do your suggested re-read there.

Code reference is here -- https://github.com/edgexfoundry/go-mod-bootstrap/blob/fe266a717da291b9ad28584cefff274bcef081c0/bootstrap/handlers/secret/secret.go#L64.  

I'd ask you to actually work the follow-on request.  Either I or @tingyuz can work with you to get it done but I'd ask that you take the lead and request help from me or him as you need it.

Reply:
Sounds good, I'll create that issue in go-mod-bootstrap today and will ping you / TingYu as needed.

Reply:
Issue created on go-mod-boostrap https://github.com/edgexfoundry/go-mod-bootstrap/issues/35.  Resolved this.